python: Switch to pyproject.toml

youtux opened the pull request:
Fixes https://github.com/cucumber/gherkin/issues/289

Notes:
I had to remove the `bin/gherkin` executable since that doesn't work with pyproject.toml (scripts can only be python entrypoints). But I replaced with the correct annotation in pyproject.toml, so if you install the package, the script will be automatically installed in the venv too

I uploaded the package built using `python -m build` to https://test.pypi.org/project/youtux.gherkin-official/, in case somebody wants to check how it will look on PyPI or to try it out.

<!---
Thanks for helping to make Cucumber better! 💖

You can feel free to open a "Draft" status pull request if you're not ready for feedback yet.

Don't worry about getting everything perfect! We're here to help and will coach you through
to getting your pull request ready to merge.

The prompts below are for guidance to help you describe your change in a way that is most 
likely to make sense to other people when they are reviewing it. Still, it's just a guide, 
so feel free to delete anything that doesn't feel appropriate, and add anything additional 
that seems like it would provide useful context. 👏🏻
-->

### 🤔 What's changed?

<!-- Describe your changes in detail -->

### ⚡️ What's your motivation? 

<!-- 
What motivated you to propose this change? Does it fix a bug? Add a new feature?
If it fixes an open issue, you can link to the issue here, e.g. "Fixes #99"
-->

### 🏷️ What kind of change is this?

<!--- Delete any options that are not relevant -->

- :bank: Refactoring/debt/DX (improvement to code design, tooling, etc. without changing behaviour)

### ♻️ Anything particular you want feedback on?

<!-- 
Is there anything in this change you're unsure about, or would 
particularly like reviewers to give you feedback on?
-->

### 📋 Checklist:

<!--- 
This is to help you remember all the little things we often forget to do!

Feel free to delete any tasks that are not relevant, or add new ones.
-->

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://github.com/cucumber/.github/tree/main?tab=coc-ov-file)
- [ ] I've changed the behaviour of the code
  - [ ] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [ ] Users should know about my change
  - [ ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request.

----

*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

jenisys commented:
Just a comment:
* You did not need to remove the `bin/gherkin` shell script.
* `pyproject.toml` just needs to generate a python-script instead (which you did).
* If it makes sense and is of any use for the local developer, keep the original `bin/gherkin` script or provide a simple python script (equivalent what is generated by install this package).

jenisys commented:
RELATED TO: `pyproject.toml`
* Line: `scripts = { gherkin = "gherkin" }` is probably wrong I think.
* You need a `main()` function (or another function in `gherkin.__main__.py`)
* You need to use instead (and correct the weird double-quotes below):

```
[project.scripts]
gherkin = “gherkin.__main__.main”
```

jenisys reviewed (commented):
This line is probably wrong and you need a `main()` function in `gherkin.__main__.py` (AFAIK; or any other function that you can refer to).
See comment in the comments section.

youtux commented:
So, given these scripts are only used internally by acceptance tests, I removed them from the entrypoints, so that we wont't try to install the `gherkin` executable in our users' virtualenv, since it's completely useless to them.

I adjusted our makefile accordingly to instead call `python -m gherkin.scripts.gherkin` and `python -m gherkin.scripts.generate_tokens`

youtux commented:
@jenisys  can you review again?

jenisys commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ This document is formatted according to the principles of [Keep A CHANGELOG](htt
 
 ### Removed
 - [Python] Drop compatibility for python 2. Supported python versions are 3.8, 3.9, 3.10, 3.12
+- [Python] Removed installation of `gherkin` script. It was used for internal acceptance tests only.
```
BAD IDEA: Removing the scripts from the package description
* The user of this package wants to use the scripts and not some module statements that he/she does not know
* Using `gherkin/__main__.py` should be kept because that is the natural choice for the main program
* Usability for the user sucks if you do that
* It might be OK for the repo development in Makefile but even there, it is rather suboptimal (IMHO).

youtux replied:
actually, I'm even more inclined to not include the `scripts/` folder in the distribution at all.

The reason is that these scripts are only used internally, as there is absolutely no API or documentation about these. Hell, I can't even figure out what the `gherkin` script does at all, since there is no documentation.

I really doubt that any downstream user is using these. If anything, we are just polluting their PATH with this irrelevant `gherkin` executable.

In the remote hypothesis that a downstream user was relying on this, we can always release a hotfix with the script included in the distribution.

jsa34 replied:
If the executable is needed for the purposes of the user to understand how the public API works, wouldn't this be better handled with documentation?

From what I could see, the scripts seem to be necessary (or more necessary) when we had to deal with py2 and py3 and managing which command was invoked to run the tests.  As we only use py3, we don't need a script to do this for us now?

I may have missed something, but I didn't use this executable or use it to understand how to use gherkin when I implemented it in pytest-bdd.

jenisys replied:
If these scripts/executables are only for internal use, then why put them in the Python package and not put the in `bin/` directory where they were before (at least one of them).

In addition, the `gherkin` script is provided in the current package version and before. Therefore, the script was part of the “API” of this package. By removing them you basically break this API in the next version. And you should be aware of that.

youtux replied:
> If these scripts/executables are only for internal use, then why put them in the Python package and not put the in bin/ directory where they were before (at least one of them

I can put them out of the `gherkin` package, so that they will automatically be out of the distributed package.

> In addition, the gherkin script is provided in the current package version and before. Therefore, the script was part of the “API” of this package. By removing them you basically break this API in the next version. And you should be aware of that.

yes I know, but every release is a major release it seems, so semver semantics allow this.

youtux replied:
> I can put them out of the gherkin package, so that they will automatically be out of the distributed package.

Done in b94911b060c7c1e15d4e200633a1341726cef7f6

jenisys commented on python/gherkin/scripts/generate_tokens.py:
```diff
@@ -0,0 +1,19 @@
+
+
+if __name__ == "__main__":
+    raise SystemExit(main())
```
BAD IDEA: `raise SystemExit(main())` — Raising an exception and calling the `main()` function as side-effect
* Use `sys.exit(main())` instead

youtux replied:
`sys.exit(...)` and `raise SystemExit(...)` are completely equivalent, so this is only about style.

I prefer the latter just because it avoids an import. Wdyt?

youtux replied:
> calling the main() function as side-effect

This is not really a side-effect. You may argue that also in `sys.exit(main())` it is a side effect.

jsa34 replied:
Agreed this is mainly a styling thing from what I can see, unless I am mistaken?

jenisys replied:
Actually, it is not. It is the difference between using the official API for something and **programming by side-effects**.
* Show me one example in https://docs.python.org where your idiom is used in the context of a `__main__` section
* Show me one example why packaging generates scripts the way I describe instead of how you use it

SEE ALSO:
* https://docs.python.org/3/library/__main__.html

youtux replied:
How is this programming by side effect? Where is the side effect here?

youtux replied:
anyway, I removed `raise SystemExit` in 5405b62f to make everyone happy

jenisys commented on python/gherkin/scripts/gherkin.py:
```diff
@@ -0,0 +1,56 @@
+
+
+if __name__ == "__main__":
+    raise SystemExit(main())
```
BAD IDEA: `raise SystemExit(main())` — Explanation, see above.

youtux replied:
removed in 5405b62fb4bbfea50fdb3db4ae3ac4a71616aa7f

jenisys commented on python/pyproject.toml:
```diff
@@ -1,3 +1,52 @@
+
+[project]
+name = "youtux.gherkin-official"
+version = "29.0.0"
```
OK, now (was corrected to original package name).

jenisys commented:
REVIEWED AGAIN: see the comments.

jenisys commented on python/pyproject.toml:
```diff
@@ -1,3 +1,51 @@
+gherkin = ["gherkin-languages.json"]
+
+[tool.setuptools.packages.find]
+include = ["gherkin", "gherkin.pickles", "gherkin.stream"]
```
BETTER: Use the `find` mechanism that is described in [setuptools docs: Find simple packages](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#finding-simple-packages)
* BLOCKER: In your case the `gherkin/scripts/` directory would have been missing.

youtux replied:
actually, I'm even more inclined to not include the `scripts/` folder in the distribution at all.

The reason is that these scripts are only used internally, as there is absolutely no API or documentation about these. Hell, I can't even figure out what the `gherkin` script does at all, since there is no documentation.

I really doubt that any downstream user is using these. If anything, we are just polluting their PATH with this irrelevant `gherkin` executable.

In the remote hypothesis that a downstream user was relying on this, we can always release a hotfix with the script included in the distribution.

youtux replied:
done in d992ac46

jenisys reviewed (changes requested):
See comments

jenisys commented:
I looked over the current state again.
Unfortunately I stumbled over one major problem (and a few minors):

MAJOR:

- [ ] The "build"/installed Python package does not contain `gherkin/gherkin-languages.json` file. Therefore, internationalisation for Gherkin keywords will not work when the Python package is installed.

MINOR:

- [ ] OOPS: Makefile: `make clean-generate` removes your gherkin script.
- [ ] The `scripts/` directory is not a Python package and therefore needs no "__init__.py"
      BUT: If you do that -- Rename "gherkin.py" to "gherkin" to avoid name-clash with `gherkin` package (ImportErrors occur). In addition, add she-bang `#!/usr/bin/env python3` and executable flags. ALTERNATIVE: Move `scrips/` directory back into the package.

- [ ] OLD "MANIFEST" file is checked into repository -> SHOULD BE: Removed.

youtux commented:
> The "build"/installed Python package does not contain gherkin/gherkin-languages.json file. Therefore, internationalisation for Gherkin keywords will not work when the Python package is installed.

it does:
```
$ python -m build
❯ python -m build                                                                          
* Creating isolated environment: venv+pip...
* Installing packages in isolated environment:
  - setuptools>=61.0
  - wheel
* Getting build dependencies for sdist...
running egg_info
writing gherkin_official.egg-info/PKG-INFO
writing dependency_links to gherkin_official.egg-info/dependency_links.txt
writing requirements to gherkin_official.egg-info/requires.txt
writing top-level names to gherkin_official.egg-info/top_level.txt
reading manifest file 'gherkin_official.egg-info/SOURCES.txt'
adding license file 'LICENSE'
writing manifest file 'gherkin_official.egg-info/SOURCES.txt'
* Building sdist...
running sdist
running egg_info
writing gherkin_official.egg-info/PKG-INFO
writing dependency_links to gherkin_official.egg-info/dependency_links.txt
writing requirements to gherkin_official.egg-info/requires.txt
writing top-level names to gherkin_official.egg-info/top_level.txt
reading manifest file 'gherkin_official.egg-info/SOURCES.txt'
adding license file 'LICENSE'
writing manifest file 'gherkin_official.egg-info/SOURCES.txt'
running check
creating gherkin_official-29.0.0
creating gherkin_official-29.0.0/gherkin
creating gherkin_official-29.0.0/gherkin/pickles
creating gherkin_official-29.0.0/gherkin/stream
creating gherkin_official-29.0.0/gherkin/stream/asd
creating gherkin_official-29.0.0/gherkin_official.egg-info
copying files to gherkin_official-29.0.0...
copying LICENSE -> gherkin_official-29.0.0
copying README.md -> gherkin_official-29.0.0
copying pyproject.toml -> gherkin_official-29.0.0
copying gherkin/__init__.py -> gherkin_official-29.0.0/gherkin
copying gherkin/ast_builder.py -> gherkin_official-29.0.0/gherkin
copying gherkin/ast_node.py -> gherkin_official-29.0.0/gherkin
copying gherkin/dialect.py -> gherkin_official-29.0.0/gherkin
copying gherkin/errors.py -> gherkin_official-29.0.0/gherkin
copying gherkin/gherkin-languages.json -> gherkin_official-29.0.0/gherkin
copying gherkin/gherkin_line.py -> gherkin_official-29.0.0/gherkin
copying gherkin/inout.py -> gherkin_official-29.0.0/gherkin
copying gherkin/parser.py -> gherkin_official-29.0.0/gherkin
copying gherkin/parser_types.py -> gherkin_official-29.0.0/gherkin
copying gherkin/py.typed -> gherkin_official-29.0.0/gherkin
copying gherkin/token.py -> gherkin_official-29.0.0/gherkin
copying gherkin/token_formatter_builder.py -> gherkin_official-29.0.0/gherkin
copying gherkin/token_matcher.py -> gherkin_official-29.0.0/gherkin
copying gherkin/token_matcher_markdown.py -> gherkin_official-29.0.0/gherkin
copying gherkin/token_scanner.py -> gherkin_official-29.0.0/gherkin
copying gherkin/pickles/__init__.py -> gherkin_official-29.0.0/gherkin/pickles
copying gherkin/pickles/compiler.py -> gherkin_official-29.0.0/gherkin/pickles
copying gherkin/stream/__init__.py -> gherkin_official-29.0.0/gherkin/stream
copying gherkin/stream/gherkin_events.py -> gherkin_official-29.0.0/gherkin/stream
copying gherkin/stream/id_generator.py -> gherkin_official-29.0.0/gherkin/stream
copying gherkin/stream/source_events.py -> gherkin_official-29.0.0/gherkin/stream
copying gherkin/stream/asd/__init__.py -> gherkin_official-29.0.0/gherkin/stream/asd
copying gherkin_official.egg-info/PKG-INFO -> gherkin_official-29.0.0/gherkin_official.egg-info
copying gherkin_official.egg-info/SOURCES.txt -> gherkin_official-29.0.0/gherkin_official.egg-info
copying gherkin_official.egg-info/dependency_links.txt -> gherkin_official-29.0.0/gherkin_official.egg-info
copying gherkin_official.egg-info/requires.txt -> gherkin_official-29.0.0/gherkin_official.egg-info
copying gherkin_official.egg-info/top_level.txt -> gherkin_official-29.0.0/gherkin_official.egg-info
copying gherkin_official.egg-info/SOURCES.txt -> gherkin_official-29.0.0/gherkin_official.egg-info
Writing gherkin_official-29.0.0/setup.cfg
Creating tar archive
removing 'gherkin_official-29.0.0' (and everything under it)
* Building wheel from sdist
* Creating isolated environment: venv+pip...
* Installing packages in isolated environment:
  - setuptools>=61.0
  - wheel
* Getting build dependencies for wheel...
running egg_info
writing gherkin_official.egg-info/PKG-INFO
writing dependency_links to gherkin_official.egg-info/dependency_links.txt
writing requirements to gherkin_official.egg-info/requires.txt
writing top-level names to gherkin_official.egg-info/top_level.txt
reading manifest file 'gherkin_official.egg-info/SOURCES.txt'
adding license file 'LICENSE'
writing manifest file 'gherkin_official.egg-info/SOURCES.txt'
* Building wheel...
running bdist_wheel
running build
running build_py
creating build/lib/gherkin
copying gherkin/gherkin_line.py -> build/lib/gherkin
copying gherkin/ast_builder.py -> build/lib/gherkin
copying gherkin/token.py -> build/lib/gherkin
copying gherkin/dialect.py -> build/lib/gherkin
copying gherkin/__init__.py -> build/lib/gherkin
copying gherkin/token_matcher_markdown.py -> build/lib/gherkin
copying gherkin/parser.py -> build/lib/gherkin
copying gherkin/token_formatter_builder.py -> build/lib/gherkin
copying gherkin/parser_types.py -> build/lib/gherkin
copying gherkin/errors.py -> build/lib/gherkin
copying gherkin/token_scanner.py -> build/lib/gherkin
copying gherkin/inout.py -> build/lib/gherkin
copying gherkin/ast_node.py -> build/lib/gherkin
copying gherkin/token_matcher.py -> build/lib/gherkin
creating build/lib/gherkin/pickles
copying gherkin/pickles/compiler.py -> build/lib/gherkin/pickles
copying gherkin/pickles/__init__.py -> build/lib/gherkin/pickles
creating build/lib/gherkin/stream
copying gherkin/stream/id_generator.py -> build/lib/gherkin/stream
copying gherkin/stream/__init__.py -> build/lib/gherkin/stream
copying gherkin/stream/gherkin_events.py -> build/lib/gherkin/stream
copying gherkin/stream/source_events.py -> build/lib/gherkin/stream
creating build/lib/gherkin/stream/asd
copying gherkin/stream/asd/__init__.py -> build/lib/gherkin/stream/asd
running egg_info
writing gherkin_official.egg-info/PKG-INFO
writing dependency_links to gherkin_official.egg-info/dependency_links.txt
writing requirements to gherkin_official.egg-info/requires.txt
writing top-level names to gherkin_official.egg-info/top_level.txt
reading manifest file 'gherkin_official.egg-info/SOURCES.txt'
adding license file 'LICENSE'
writing manifest file 'gherkin_official.egg-info/SOURCES.txt'
copying gherkin/gherkin-languages.json -> build/lib/gherkin
copying gherkin/py.typed -> build/lib/gherkin
installing to build/bdist.macosx-14.6-arm64/wheel
running install
running install_lib
creating build/bdist.macosx-14.6-arm64/wheel
creating build/bdist.macosx-14.6-arm64/wheel/gherkin
copying build/lib/gherkin/gherkin_line.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/ast_builder.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/token.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/dialect.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
creating build/bdist.macosx-14.6-arm64/wheel/gherkin/pickles
copying build/lib/gherkin/pickles/compiler.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin/pickles
copying build/lib/gherkin/pickles/__init__.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin/pickles
creating build/bdist.macosx-14.6-arm64/wheel/gherkin/stream
copying build/lib/gherkin/stream/id_generator.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin/stream
creating build/bdist.macosx-14.6-arm64/wheel/gherkin/stream/asd
copying build/lib/gherkin/stream/asd/__init__.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin/stream/asd
copying build/lib/gherkin/stream/__init__.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin/stream
copying build/lib/gherkin/stream/gherkin_events.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin/stream
copying build/lib/gherkin/stream/source_events.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin/stream
copying build/lib/gherkin/__init__.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/token_matcher_markdown.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/parser.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/token_formatter_builder.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/parser_types.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/py.typed -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/gherkin-languages.json -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/errors.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/token_scanner.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/inout.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/ast_node.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
copying build/lib/gherkin/token_matcher.py -> build/bdist.macosx-14.6-arm64/wheel/./gherkin
running install_egg_info
Copying gherkin_official.egg-info to build/bdist.macosx-14.6-arm64/wheel/./gherkin_official-29.0.0-py3.8.egg-info
running install_scripts
creating build/bdist.macosx-14.6-arm64/wheel/gherkin_official-29.0.0.dist-info/WHEEL
creating '/Users/youtux/Developer/gherkin/python/dist/.tmp-tzhl9qus/gherkin_official-29.0.0-py3-none-any.whl' and adding 'build/bdist.macosx-14.6-arm64/wheel' to it
adding 'gherkin/__init__.py'
adding 'gherkin/ast_builder.py'
adding 'gherkin/ast_node.py'
adding 'gherkin/dialect.py'
adding 'gherkin/errors.py'
adding 'gherkin/gherkin-languages.json'
adding 'gherkin/gherkin_line.py'
adding 'gherkin/inout.py'
adding 'gherkin/parser.py'
adding 'gherkin/parser_types.py'
adding 'gherkin/py.typed'
adding 'gherkin/token.py'
adding 'gherkin/token_formatter_builder.py'
adding 'gherkin/token_matcher.py'
adding 'gherkin/token_matcher_markdown.py'
adding 'gherkin/token_scanner.py'
adding 'gherkin/pickles/__init__.py'
adding 'gherkin/pickles/compiler.py'
adding 'gherkin/stream/__init__.py'
adding 'gherkin/stream/gherkin_events.py'
adding 'gherkin/stream/id_generator.py'
adding 'gherkin/stream/source_events.py'
adding 'gherkin/stream/asd/__init__.py'
adding 'gherkin_official-29.0.0.dist-info/LICENSE'
adding 'gherkin_official-29.0.0.dist-info/METADATA'
adding 'gherkin_official-29.0.0.dist-info/WHEEL'
adding 'gherkin_official-29.0.0.dist-info/top_level.txt'
adding 'gherkin_official-29.0.0.dist-info/RECORD'
removing build/bdist.macosx-14.6-arm64/wheel
Successfully built gherkin_official-29.0.0.tar.gz and gherkin_official-29.0.0-py3-none-any.whl

```

youtux commented:
> OLD "MANIFEST" file is checked into repository -> SHOULD BE: Removed (clean up prior problem).

good one, deleted in 987e836d05cd6934c48fe7c61abc06b31ad1114d

youtux commented:
> OOPS: Makefile: make clean-generate removes your gherkin script.

It always did. That's the expected behaviour.

youtux commented:
> The scripts/ directory is not a Python package and therefore needs no __init__.py.
BUT: If you do that -- Rename "gherkin.py" to "gherkin" to avoid name-clash with gherkin package (ImportErrors occur). In addition, add she-bang #!/usr/bin/env python3 and executable flags. ALTERNATIVE: Move scrips/ directory back into the package.

Removed `__init__.py`. It actually didn't cause any import issue. But to be safe, I renamed it to `generate_events` to be consistent with the other file. 3ed84b3d

youtux commented:
@jenisys please review again

jenisys commented:
## RELATED TO: `gherkin-languages.json`

I must have been mistaken yesterday:

* I build the sdist package and unpacked it in another directory
* I was trying to use tests and the scripts and ran into a `FileNotFoundError` related to `gherkin-languages.json`

Everything seems to be OK now.

## Makefile

After you removed `scripts/__init__.py`, you should basically use `python $SCRIPT_PATH` instead of `python -m $SCRIPT_MODULE` in the `Makefile`.
Weird that the module-syntax seems to be working if you run `make acceptance`.

```make
# -- FILE: python/Makefile
...
# ORIGINAL: GHERKIN_GENERATE_EVENTS = python -m scripts.generate_events
# ORIGINAL: GHERKIN_GENERATE_TOKENS = python -m scripts.generate_tokens
GHERKIN_GENERATE_EVENTS = python scripts/generate_events.py
GHERKIN_GENERATE_TOKENS = python scripts/generate_tokens.py
...
```

## Python Source-Code Package (`sdist`)

Normally, the source code package (aka: `sdist`) as TARGZ/ZIP archive should contain the complete contents of the source code repository (without the VCS details and temporary directory/files). If this is the case, you can unpack this archive and run the tests in your workspace (for example).

Currently, the following directories/files are missing:

* `test/` subdirectory
* Makefile (questionable if this should be included because `../testdata/` is missing)
* `gherkin-python.razor` (questionable if this should be included because `../gherkin-berp` is missing)

I would include the `test/` directory at least.
Think about it and make a decision.

## Otherwise

* `gherkin-python.razor` has executable permissions (same as in the other directories, like: `../javascript/gherkin-javascript.razor`). This is not needed (IMHO). But such a clean up should occur for all languages.

jenisys reviewed (commented):
DONE: See comments.

youtux commented:
> Weird that the module-syntax seems to be working if you run make acceptance.

Why is it weird? `scripts.generate_tokens` is a valid module name from the cwd.

> Normally, the source code package (aka: sdist) as TARGZ/ZIP archive should contain the complete contents of the source code repository

An source for this? To me, the sdist should just contain enough info to build the package, not to test it. If we want to make the sdist testable, we should include all the testdata from upper level too. Otherwise it would just be half done.

jenisys commented:
MERGED INTO: main (manually and squashed commits, shortened commit-message)
